File: src/lexer/token.ts
import type { Location } from "../diagnostic/location";
import { ParseError } from "../parser/parser";

export const TokenType = {
    EndOfInput: { identifier: "EndOfInput" },
    Identifier: { identifier: "Identifier" },
    Whitespace: { identifier: "Whitespace" },

    Function: { identifier: "Function", value: "func" },
    For: { identifier: "For", value: "for" },
    ForEach: { identifier: "ForEach", value: "foreach" },
    If: { identifier: "If", value: "if" },
    While: { identifier: "While", value: "while" },
    StringType: { identifier: "StringType", value: "string" },
    NumberType: { identifier: "NumberType", value: "number" },
    BooleanType: { identifier: "BooleanType", value: "boolean" },
    VoidType: { identifier: "VoidType", value: "void" },
    AnyType: { identifier: "AnyType", value: "any" },
    Return: { identifier: "Return", value: "return" },
    Continue: { identifier: "Continue", value: "continue" },
    Break: { identifier: "Break", value: "break" },

    Number: { identifier: "Number" },
    Boolean: { identifier: "Boolean" },
    String: { identifier: "String" },

    OpenParen: { identifier: "OpenParen", value: "(" },
    CloseParen: { identifier: "CloseParen", value: ")" },
    OpenBracket: { identifier: "OpenBracket", value: "[" },
    CloseBracket: { identifier: "CloseBracket", value: "]" },
    OpenBrace: { identifier: "OpenBrace", value: "{" },
    CloseBrace: { identifier: "CloseBrace", value: "}" },
    And: { identifier: "And", value: "&&" },
    Or: { identifier: "Or", value: "||" },
    Colon: { identifier: "Colon", value: ":" },
    Comma: { identifier: "Comma", value: "," },
    Period: { identifier: "Period", value: "." },
    Arrow: { identifier: "Arrow", value: "->" },
    Equals: { identifier: "Equals", value: "==" },
    Assign: { identifier: "Assign", value: "=" },
    NotEquals: { identifier: "NotEquals", value: "!=" },
    GreaterEquals: { identifier: "GreaterEquals", value: ">=" },
    LessEquals: { identifier: "LessEquals", value: "<=" },
    GreaterThan: { identifier: "GreaterThan", value: ">" },
    LessThan: { identifier: "LessThan", value: "<" },
    PlusAssign: { identifier: "PlusAssign", value: "+=" },
    MinusAssign: { identifier: "MinusAssign", value: "-=" },
    MultiplyAssign: { identifier: "MultiplyAssign", value: "*=" },
    DivideAssign: { identifier: "DivideAssign", value: "/=" },
    Increment: { identifier: "Increment", value: "++" },
    Decrement: { identifier: "Decrement", value: "--" },
    Plus: { identifier: "Plus", value: "+" },
    Minus: { identifier: "Minus", value: "-" },
    Multiply: { identifier: "Multiply", value: "*" },
    Divide: { identifier: "Divide", value: "/" },
}

export type TokenTypeData = { identifier: string, value?: string };

export class Token<T = null> {
    constructor(
        public type: keyof typeof TokenType,
        public raw: string,
        public location: Location,
        public value: T
    ) {}

    public expect(
        type: keyof typeof TokenType,
        message: string = "Expected '{0}', found '{1}'"
    ): this {
        const expected = typeAsString(type);
        const actual = this.asString();

        if (this.type !== type)
            throw new ParseError(message.replace(/\{0}/g, expected)
                .replace(/\{1}/g, actual));

        return this;
    }

    public asString(): string {
        switch (this.type) {
            case "Boolean":
            case "Number": {
                return (this.value as any).toString();
            }
        }

        const data = TokenType[this.type] as TokenTypeData;
        return data.value ?? this.type;
    }
}

export function typeAsString(type: keyof typeof TokenType) {
    const data = TokenType[type] as TokenTypeData;
    return data.value ?? type;
}

File: src/main/kotlin/me/honkling/miniscript/lexer/Token.kt
package me.honkling.miniscript.lexer

import me.honkling.miniscript.diagnostic.Location
import me.honkling.miniscript.diagnostic.Logger
import me.honkling.miniscript.diagnostic.MiniScriptException

open class Token(
    val type: TokenType,
    val raw: String,
    val location: Location,
    private val logger: Logger
) {
    class WithValue<T : Any>(
        type: TokenType,
        raw: String,
        location: Location,
        logger: Logger,
        val value: T
    ) : Token(type, raw, location, logger)

    fun expect(
        type: TokenType,
        message: String = "Expected {0}, found {1}"
    ): Token {
        if (this.type != type) {
            logger.error(message.replace("{0}", asString(type))
                .replace("{1}", asString()))

            throw MiniScriptException.ParseError()
        }

        return this
    }

    fun asString(type: TokenType = this.type): String {
        val value = type.value?.let { "'$it'" } ?: type.name.lowercase()

        if (this.type != type)
            return value

        return when (type) {
            TokenType.Boolean, TokenType.Number -> "'${(this as Token.WithValue<*>).value}'"
            TokenType.EndOfInput -> "end of file"
            else -> value
        }
    }

    override fun toString(): String {
        return "Token(type=${type.name}, raw=\"${raw.replace("\"", "\\\"")}\", location=$location)"
    }
}

enum class TokenType(val value: String? = null) {
    EndOfInput,
    Identifier,
    Whitespace,

    Function("func"),
    For("for"),
    ForEach("foreach"),
    If("if"),
    Else("else"),
    While("while"),
    StringType("string"),
    NumberType("number"),
    BooleanType("boolean"),
    DictType("dict"),
    VoidType("void"),
    AnyType("any"),
    Return("return"),
    Continue("continue"),
    Break("break"),
    Native("native"),

    Number,
    Boolean,
    String,

    OpenParen("("),
    CloseParen(")"),
    OpenBracket("["),
    CloseBracket("]"),
    OpenBrace("{"),
    CloseBrace("}"),
    And("&&"),
    Or("||"),
    Colon(":"),
    Comma(","),
    Period("."),
    Arrow("->"),
    Equals("=="),
    Assign("="),
    NotEquals("!="),
    GreaterEquals(">="),
    LessEquals("<="),
    GreaterThan(">"),
    LessThan("<"),
    PlusAssign("+="),
    MinusAssign("-="),
    MultiplyAssign("*="),
    DivideAssign("/="),
    Increment("++"),
    Decrement("--"),
    Plus("+"),
    Minus("-"),
    Multiply("*"),
    Divide("/")
}